### PR TITLE
nixos/k3s: streamline HA setup

### LINKED ISSUE #182085: k3s: support HA cluster
### Describe the bug
While trying to set up a NixOS HA k3s cluster, I found this: https://github.com/NixOS/nixpkgs/blob/c8585bf4a88d6ec2b659cf5fd02a10cac2d31c82/nixos/modules/services/cluster/k3s/default.nix#L34-L55

Notice how it says that those options only make sense for an agent? [It's not true](https://rancher.com/docs/k3s/latest/en/installation/ha-embedded/). Otherwise, how are we supposed to build a HA cluster? (With 3 or more servers)

### Expected behavior
There should be a documented path towards building an HA cluster.

### Additional context
Embedded etcd is these days the default storage endpoint for k3s. This means that HA clusters are supported out of the box.

For NixOS, all I had to do is deploy this module on all servers:

```nix
firstServer: firstServerEndpoint: host: {
  services.k3s.enable = true;
  environment.systemPackages = [pkgs.k3s];
  services.k3s.extraFlags = "--secrets-encryption ${
    if host == firstServer
    then "--cluster-init"
    else ""
  }";
  networking.firewall.allowedTCPPorts = [6443];

  # These work fine for servers, no matter what option docs say
  services.k3s.token = "mySecretToken";
  services.k3s.serverAddr =
    if host == firstServer
    then ""
    else firstServerEndpoint;

  # This is important for server nodes to reach each other
  networking.firewall.allowedTCPPortRanges = [
    {
      from = 2379;
      to = 2380;
    }
  ];
}
```

Another key point is [the etcd port range that has to be open for multiple servers to join together](https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/#networking). [The wiki](https://nixos.wiki/wiki/K3s) doesn't mention that.

Finally, if you're deploying in parallel (NixOps or similar), not-first nodes (agents or servers) can have a race condition waiting for the 1st server to be available. It'd be nice to be able to configure those services to wait automatically for some kind of socket availability (maybe defined as a systemd unit?).

### Notify maintainers

@euank @mic92 @superherointj

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.54, NixOS, 22.05 (Quokka), 22.05.20220710.cf034a8`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - channels(root): `"nixos-22.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

@moduon MT-904

### PR DESCRIPTION
**Pull request by yajo**
###### Description of changes

- Remove fake docs.
- Add new assertions to let configurations make more sense.
- Add clusterInit flag.
- Add some more docs about HA mode setup.

Fix https://github.com/NixOS/nixpkgs/issues/182085

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@moduon MT-904

### CODE REVIEW

**Comment by superherointj** [comment]
Please follow:

https://nixos.org/manual/nixos/unstable/#sec-writing-documentation

**Comment by superherointj** [comment]
Notice CI errors on docs. Please follow:

https://nixos.org/manual/nixos/unstable/#sec-writing-documentation

File: nixos/modules/services/cluster/k3s/default.nix [hunk 1]
```diff
@@ -33,15 +33,38 @@ in
...
+        [1]: https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/#networking
+      '';
       example = "https://10.0.0.10:6443";
       default = "";
     };
 
+    clusterInit = mkOption {
+      type = types.bool;
+      default = cfg.role == "server";
+      example = false;
```
Comment (superherointj) [inline nixos/modules/services/cluster/k3s/default.nix:51]:
```suggestion
      defaultText = "server";
      example = false;
```

**Comment by superherointj** [comment]
`Remove fake docs.` is bad wording.
Docs aren't fake. And you aren't removing docs. Please improve on it.

**Comment by superherointj** [comment]
Would be nice if you also contributed a test to test the HA feature you are adding. Feel free to take ideas from [here](https://github.com/NixOS/nixpkgs/tree/master/nixos/tests/k3s).

File: nixos/modules/services/cluster/k3s/default.nix [hunk 2]
```diff
@@ -33,15 +33,38 @@ in
...
     serverAddr = mkOption {
       type = types.str;
-      description = lib.mdDoc "The k3s server to connect to. This option only makes sense for an agent.";
+      description = lib.mdDoc ''
+        The k3s server to connect to.
+
+        Servers and agents need to communicate each other.
+        Read [the networking docs][1] to know how to configure the firewall.
+
+        [1]: https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/#networking
```
Comment (pennae) [inline nixos/modules/services/cluster/k3s/default.nix:42]: individual descriptions are not independent markdown instances, using references like this could lead to problems. (this is the first one to appear at all)

Comment (superherointj) [reply nixos/modules/services/cluster/k3s/default.nix:42]: Maybe we could have an option to open firewall ports for K3s.

**Comment by superherointj** [comment]
Please fix:
image

File: nixos/modules/services/cluster/k3s/default.nix (hunk 2 above)
Comment (Mic92) [inline nixos/modules/services/cluster/k3s/default.nix:42]:
```suggestion
        Read (the networking docs)[https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/#networking] to know how to configure the firewall.
```

Comment (hall) [reply nixos/modules/services/cluster/k3s/default.nix:42]: I'd like to see an `openFirewall` option here. If it helps, I've just recently setup an HA k3s cluster and these are the ports which had to be configured:
```nix
{
  networking.firewall = {
    allowedUDPPorts = [
      53 # dns
      8472 # cni (flannel vxlan)
    ];
    allowedTCPPorts = [
      10250 # metrics server
    ] ++ (lib.optionals (cfg.role == "server") [
      6443 # k8s api
      2379 # etcd client requests
      2380 # etcd peer communication
    ]);
  };
}
```

**Comment by superherointj** [comment]
@yajo Please update the PR. And report status.

File: nixos/modules/services/cluster/k3s/default.nix (hunk 2 above)
Comment (yajo) [reply nixos/modules/services/cluster/k3s/default.nix]: I deliberately didn't add that option because it depends a lot on how you want to configure the firewall.

The linked docs explain that some ports you opened shouldn't be world-reachable. Instead, most ports should be only reachable among nodes. ⚠️ Beware if that's the setup you're deploying in production!

To properly set up the firewall connections, we need to know other nodes' IPs. That fits better in multi-node deployment tools such as nixops or similar.

So, for now, I'll leave the firewall setup to the sysadmin.

**Review by yajo — COMMENTED** [review]
Sorry for the late reply, I was on holidays.

All comments attended, except for the tests. I'll look into them now.

**Comment by yajo** [comment]
Tests added.

File: nixos/tests/k3s/multi-node.nix [hunk 3]
```diff
@@ -65,14 +65,38 @@ import ../make-test-python.nix ({ pkgs, ... }:
...
+      server2 = { pkgs, ... }: {
+        environment.systemPackages = with pkgs; [ gzip jq ];
+        virtualisation.memorySize = 1536;
+        virtualisation.diskSize = 4096;
+
+        services.k3s = {
+          inherit tokenFile;
+          enable = true;
+          role = "server";
+          package = pkgs.k3s;
```
Comment (superherointj) [inline nixos/tests/k3s/multi-node.nix:77]:
```suggestion
```
Already default values.

File: nixos/tests/k3s/multi-node.nix [hunk 4]
```diff
@@ -65,14 +65,38 @@ import ../make-test-python.nix ({ pkgs, ... }:
...
+          serverAddr = "https://192.168.1.1:6443";
+          clusterInit = false;
+          extraFlags = "--no-deploy coredns,servicelb,traefik,local-storage,metrics-server --pause-image test.local/pause:local --node-ip 192.168.1.3";
+        };
+        networking.firewall.allowedTCPPorts = [ 2379 2380 6443 ];
+        networking.firewall.allowedUDPPorts = [ 8472 ];
+        networking.firewall.trustedInterfaces = [ "flannel.1" ];
+        networking.useDHCP = false;
+        networking.defaultGateway = "192.168.1.3";
```
Comment (superherointj) [inline nixos/tests/k3s/multi-node.nix:86]:
```suggestion
        networking.defaultGateway = "192.168.1.3";
```
Why gateway is itself? In case this is not needed. It shouldn't be included then?

File: nixos/modules/services/cluster/k3s/default.nix [hunk 5]
```diff
@@ -33,15 +33,40 @@ in
...
+        [the networking docs](https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/#networking)
+        to know how to configure the firewall.
+      '';
       example = "https://10.0.0.10:6443";
       default = "";
     };
 
+    clusterInit = mkOption {
+      type = types.bool;
+      default = cfg.role == "server";
```
Comment (superherointj) [inline nixos/modules/services/cluster/k3s/default.nix:49]:
```suggestion
      default = cfg.role == "server" && cfg.serverAddr == "";
```

File: nixos/modules/services/cluster/k3s/default.nix [hunk 6]
```diff
@@ -33,15 +33,40 @@ in
...
+      defaultText = literalExpression ''cfg.role == "server"'';
```
Comment (superherointj) [inline nixos/modules/services/cluster/k3s/default.nix:50]:
```suggestion
      defaultText = ''false";
```
Why is it necessary otherwise?

File: nixos/modules/services/cluster/k3s/default.nix (hunk 5 above)
Comment (superherointj) [reply nixos/modules/services/cluster/k3s/default.nix:49]: It seems we have 3 roles: `master`, `replica`, `node`.
"master" won't have `serverAddr` but will have `clusterInit` true.
`replica` is a server but is not master, requires serverAddr and denies use of clusterInit.
Maybe we should change roles to reflect this automatically. And infer from there. Also warn when some field is missing.
`nodes` seek to use all IPs from master & replica in serverAddr.
Do you agree?

Comment (superherointj) [reply nixos/modules/services/cluster/k3s/default.nix:49]: @euank Do you think `roles` option should account for HA needs?

Comment (euank) [reply nixos/modules/services/cluster/k3s/default.nix:49]: I'm not totally sure.

The "roles" thing is there because the upstream docs and installation instructions talk a lot about "agent" vs "server".

Ideally, I'd like to keep our concepts as close to the upstream k3s docs so that those docs can help someone figure out how to use the nixos module too.

For consistency with [upstream docs](https://rancher.com/docs/k3s/latest/en/installation/ha-embedded/), I'd prefer to expose `clusterInit` over adding a new replica role to mask over it.

File: nixos/modules/services/cluster/k3s/default.nix [hunk 7]
```diff
@@ -33,15 +33,40 @@ in
...
+      example = false;
+      description = lib.mdDoc ''
+        Initialize a cluster if k3s is started for the first time.
+
+        If the cluster is already initialized, this option has no effect.
+
+        This option only makes sense in a server that is not connecting to another server.
+
+        If you are configuring an HA cluster, the 1st server must be configured
```
Comment (euank) [inline nixos/modules/services/cluster/k3s/default.nix:59]: "an HA cluster" to "an HA cluster with an embedded DB" maybe? I know the doc comment's also already a bit wordy, but the other type of HA (external datastore) doesn't use the cluster-init flag at all.

File: nixos/modules/services/cluster/k3s/default.nix [hunk 8]
```diff
@@ -33,15 +33,40 @@ in
...
 
+    clusterInit = mkOption {
+      type = types.bool;
+      default = cfg.role == "server";
+      defaultText = literalExpression ''cfg.role == "server"'';
+      example = false;
+      description = lib.mdDoc ''
+        Initialize a cluster if k3s is started for the first time.
+
+        If the cluster is already initialized, this option has no effect.
```
Comment (euank) [inline nixos/modules/services/cluster/k3s/default.nix:55]: From [the docs](https://rancher.com/docs/k3s/latest/en/installation/ha-embedded/#existing-clusters), it sounds like this does have an effect on existing clusters.

As it says:

> If you have an existing cluster using the default embedded SQLite database, you can convert it to etcd by simply restarting your K3s server with the --cluster-init flag

Since I think sqlite is / has-been the default, that means many existing clusters will be using it, and adding this default will migrate them forward to etcd.

The upstream default is still sqlite, and you have to opt-in to etcd, so I feel like that's the better default for us.

**Review by euank — COMMENTED** [review]
Thanks for the PR!

I've got a question about defaulting on `--cluster-init` being the right default, but adding the ability to get a working HA w/ embedded etcd is great!

File: nixos/modules/services/cluster/k3s/default.nix (hunk 8 above)
Comment (yajo) [reply nixos/modules/services/cluster/k3s/default.nix:55]: I can expose the `--datastore-endpoint` option. Then assert that it is `etcd` if we're enabling cluster-init.

Finally I can change the defaults to be sqlite without cluster-init, to stick closer to upstream.

Does that seem a good solution?

File: nixos/tests/k3s/multi-node.nix (hunk 4 above)
Comment (yajo) [reply nixos/tests/k3s/multi-node.nix:86]: I have no idea, I was just copying the other server definitions.

File: nixos/modules/services/cluster/k3s/default.nix (hunk 5 above)
Comment (yajo) [reply nixos/modules/services/cluster/k3s/default.nix:49]: I also think it's better to stick closer to upstream.

File: nixos/modules/services/cluster/k3s/default.nix (hunk 8 above)
Comment (yajo) [reply nixos/modules/services/cluster/k3s/default.nix:55]: Reading https://rancher.com/docs/k3s/latest/en/installation/datastore/#external-datastore-configuration-parameters I can see that `--datastore-endpoint` shall be kept empty for any embedded DB usage.

It seems like k3s is smart enough to use embedded sqlite by default and to migrate to embedded etcd automatically when the server gets `--cluster-init` or when another server joins the 1st one.

So I'll just keep this option `false` by default and improve the docs to let users understand that this flag enables embedded etcd mode.

File: nixos/tests/k3s/multi-node.nix (hunk 4 above)
Comment (yajo) [reply nixos/tests/k3s/multi-node.nix:86]: I tried to remove this and the test failed, so it was needed after all.

**Review by yajo — COMMENTED** [review]
Thanks everybody. All comments attended. I think the docs now state more clearly the different use cases.

**Comment by Mic92** [comment]
Thanks!

**Comment by superherointj** [comment]
I was not done.
